docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,678 @@
+ *        Optionally ORed with one of following page types:
+ *             - SGX_EMA_PAGE_TYPE_REG: regular page type. This is the default if not specified.
+ *             - SGX_EMA_PAGE_TYPE_SS_FIRST: the first page in shadow stack.
+ *             - SGX_EMA_PAGE_TYPE_SS_REST: the rest page in shadow stack.
+ * @retval 0 The operation was successful.
+ * @retval EINVAL Any parameter passed in is not valid.
+ * @retval errno Error as reported by dependent syscalls, e.g., mmap().
+ */
+int sgx_mmap_ocall(void *addr, size_t length, int prot, int flags);
+
+/*
+ * Call OS to change permissions, type, or notify EACCEPT done after TRIM.
+ *
+ * @param[in] addr start address of the memory to change protections.
+ * @param[in] length length of the area.  This must be a multiple of the page size.
+ * @param[in] flags this must be OR'ed of following:
+ *            SGX_EMA_PROT_READ
+ *            SGX_EMA_PROT_WRITE
+ *            SGX_EMA_PROT_EXEC
+ *            SGX_EMA_PAGE_TYPE_TRIM: change the page type to PT_TRIM, implies RW.
+ *            SGX_EMA_PAGE_TYPE_TCS: change the page type to PT_TCS
+ *            SGX_EMA_PROT_NOACCESS: Signal the kernel EACCEPT is done for PT_TRIM pages.
+ * @retval 0 The operation was successful.
+ * @retval EINVAL A parameter passed in is not valid.
+ * @retval errno Error as reported by dependent syscalls, e.g., mprotect().
+ */
+
+int sgx_mprotect_ocall(void *addr, size_t length, int flags);
+
+/*
+ * OCall to make regular munmap syscall.
+ * @param[in] addr start address of the memory to unmap.
+ * @param[in] length length of the area.  This must be a multiple of the page size.
+ * @retval 0 The operation was successful.
+ * @retval errno Error as reported by munmap syscall.
+ */
+int sgx_munmap_ocall(void *addr, size_t length);
+
+
+```
+
+### Other Utilities
+
+```
+/*
+ * Define a mutex and init/lock/unlock/destroy functions.
+ */
+typedef struct sgx_mutex_t;
+int sgx_mutex_init(sgx_mutex_t* mutex);
+int sgx_mutex_lock(sgx_mutex_t* mutex);
+int sgx_mutex_unlock(sgx_mutex_t* mutex);
+int sgx_mutex_destroy(sgx_mutex_t* mutex);
+
+/*
+ * Check whether the given buffer is strictly within the enclave.
+ *
+ * Check whether the buffer given by the **ptr** and **size** parameters is
+ * strictly within the enclave's memory. If so, return true. If any
+ * portion of the buffer lies outside the enclave's memory, return false.
+ *
+ * @param[in] ptr The pointer pointer to buffer.
+ * @param[in] size The size of the buffer
+ *
+ * @retval true The buffer is strictly within the enclave.
+ * @retval false At least some part of the buffer is outside the enclave, or
+ * the arguments are invalid. For example, if **ptr** is null or **size**
+ * causes arithmetic operations to wrap.
+ *
+ */
+bool sgx_is_within_enclave(const void *ptr, size_t size);
+
+```
+
+### Support for EMM Initialization
+
+In addition to implement the abstraction layer APIs, a runtime shall provide
```

Comment:
The initial idea was for runtime to call ema_alloc with EMA_SYSTEM flag to set up those initial layouts. And other private APIs below can be used to adjust attributes and boundaries of initially committed pages

Now I am also thinking to add one more API specific for EADDed regions, some thing like this:
```
/*
 * Initialize an EMA. This can be used to setup EMAs to account regions that
 * are loaded and initialized with EADD before EINIT.  
 * @param[in] addr Starting address of the region, page aligned. If NULL is provided,
 *                  then the function will select the starting address.
 * @param[in] length Size of the region in multiples of page size in bytes.
 * @param[in] flags SGX_EMA_SYSTEM, or SGX_EMA_SYSTEM | SGX_EMA_RESERVE
 *           bitwise ORed with one of following page types:
 *             - SGX_EMA_PAGE_TYPE_REG: regular page type. This is the default if not specified.
 *             - SGX_EMA_PAGE_TYPE_TCS: TCS page.
 *             - SGX_EMA_PAGE_TYPE_SS_FIRST: the first page in shadow stack.
 *             - SGX_EMA_PAGE_TYPE_SS_REST: the rest page in shadow stack.
 * @param[in] prot permissions, either SGX_EMA_PROT_NONE or a bitwise OR of following with:
 *        - SGX_EMA_PROT_READ: Pages may be read.
 *        - SGX_EMA_PROT_WRITE: Pages may be written.
 *        - SGX_EMA_PROT_EXECUTE: Pages may be executed.
 * @param[in] handler A custom handler for page faults in this region, NULL if
 *                     no custom handling needed.
 * @param[in] handler_private Private data for the @handler, which will be passed
 *                     back when the handler is called.
 * @retval 0 The operation was successful.
 * @retval EACCES Region is outside enclave address space.
 * @retval EEXIST Any page in range requested is in use.
 * @retval EINVAL Invalid page type, flags, or addr and length are not page aligned.
 */ 
int ema_init(void *addr, size_t length, int flags, int prot,
              sgx_enclave_fault_handler_t handler,
              void *handler_private);
```